File: MyLeetCode/python/Valid Word Abbreviation.py
###
# Two Pointers
# Time Complexity: O(n)
# Space Complexity: O(1)
###
class Solution(object):
    def validWordAbbreviation(self, word, abbr):
        """
        :type word: str
        :type abbr: str
        :rtype: bool
        """
        m = len(word)
        n = len(abbr)
        i = 0
        j = 0
        while i < m and j < n:
            if abbr[j].isdigit():
                if abbr[j] == '0':
                    return False
                s = j
                while j < n and abbr[j].isdigit():
                    j += 1
                i += int(abbr[s:j])
            else:
                if word[i] != abbr[j]:
                    return False
                i += 1
                j += 1
        return i == m and j == n
        
        
        

File: MyLeetCode/Leetcode/Valid Word Abbreviation.cpp
class Solution {
public:
    bool validWordAbbreviation(string word, string abbr) {
        int n = word.size();
        int m = abbr.size();
        int ptr1 = 0;
        int ptr2 = 0;
        while (ptr2 < m && ptr1 < n) {
            if (isdigit(abbr[ptr2])) {
                if (abbr[ptr2] == '0') return false;
                int cnt = 0;
                while (ptr2 < m && isdigit(abbr[ptr2])) {
                    cnt = cnt * 10 + (abbr[ptr2] - '0');
                    ptr2++;
                }
                ptr1 += cnt;
                if (ptr1 > n) return false;
            }
            else {
                if (word[ptr1] != abbr[ptr2]) return false;
                ptr1++;
                ptr2++;
            }
        }
        return ptr1 == n && ptr2 == m;
    }
};